.github/workflows/test.yml

```diff
@@ -31,51 +29,62 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - name: Checkout
-        uses: actions/checkout@v2
+        uses: actions/checkout@v4
       - name: Setup Rust toolchain
-        uses: actions-rs/toolchain@v1
+        uses: dtolnay/rust-toolchain@master
+        id: rust-toolchain
         with:
           toolchain: ${{ matrix.rust }}
           components: rustfmt, clippy
-          default: true
-      - name: Format
-        uses: actions-rs/cargo@v1
+      - uses: actions/cache@v4
         with:
-          command: fmt
-          args: -- --check
+          path: |
+            ~/.cargo/bin/
+            ~/.cargo/registry/index/
+            ~/.cargo/registry/cache/
+            ~/.cargo/git/db/
+            target/
+          key: ${{ runner.os }}-cargo-${{ steps.rust-toolchain.outputs.cachekey }}-${{ hashFiles('**/Cargo.toml') }}
+          restore-keys: |
+            ${{ runner.os }}-cargo-${{ steps.rust-toolchain.outputs.cachekey }}-
+            ${{ runner.os }}-cargo-
```

Comment:
What's your experience for using github action's cache directly vs https://github.com/Swatinem/rust-cache. Or does this set up caching in addition to Swatinem because https://github.com/actions-rust-lang/setup-rust-toolchain enables caching by default.

Reply:
I'm using a similar setup for other crates and it works fine to reduce job times. This is not in addition to anything else, this PR proposes to use https://github.com/dtolnay/rust-toolchain/.

Checking in a `Cargo.lock` file would make this more efficient.

Reply:
I don't really have any opinion on these things. Happy to trust your judgment here.